### PR TITLE
[Open Source] Add ability to restrict channels

### PR DESCRIPTION
**Pull request by loganmeetsworld**
## Why
Sometimes we want to restrict commands so that they can only be run in certain channels. cog-consul `write` is a good example for this because it is actively changing a consul `kv`. While reading secrets is ok in `direct`, we don't want devs to do this in private. 

## What
Using dynamic configuration this is made simple. We can specify the channels we want to have for specific commands privately using `cogctl` and put a catch for those channels at the top of the `run_command`.

## Feedback
Any unforeseen use cases? 

cc/ @talaris @ktheory @natachaS

### CODE REVIEW

**Comment by loganmeetsworld** [comment]
[ChatOps/Cloudformation: require important changes to be in public rooms](https://trello.com/c/rMk9MpXr/841-chatops-cloudformation-require-important-changes-to-be-in-public-rooms)

File: lib/cog_cmd/consul/consul.rb [hunk 1]
```diff
@@ -13,4 +13,16 @@ def domain
...
+
+  def permitted_channels
+    env_var("CONSUL_CHANNELS")
+  end
+
+  def current_channel
+    env_var("COG_ROOM")
+  end
+
+  def restircted_channels_response
```
Comment (natachaS) [inline lib/cog_cmd/consul/consul.rb]: Typo in `restircted`.

File: spec/consul_spec.rb [hunk 2]
```diff
@@ -61,6 +63,40 @@
...
+          run_command(args: ['myKey'])
+        }.to raise_error(Cog::Error)
+      end
+    end
+  end
+
+  context 'Consul:Write' do 
+    let(:command_name) { 'write' }
+    describe 'writing a key' do
+      it 'should not allow you to run outside restircted_channels' do
```
Comment (natachaS) [inline spec/consul_spec.rb]: I think you meant `restricted`

**Review by natachaS — COMMENTED** [review]
Nice ! :tada:

**Review by talaris — CHANGES_REQUESTED** [review]
Right now the permitted channels is kinda forced. I think it would be good if the permitted channels was empty to skip the check.

File: lib/cog_cmd/consul/consul.rb [hunk 3]
```diff
@@ -13,4 +13,16 @@ def domain
   def consul_token
     env_var("CONSUL_TOKEN", required: true)
   end
+
+  def permitted_channels
+    env_var("CONSUL_CHANNELS")
```
Comment (ktheory) [inline lib/cog_cmd/consul/consul.rb]: @loganmeetsworld Probably want to parse this as a comma-delimited list:

```ruby
env_var('CONSUL_CHANNELS').split(/,\s*/)
```

As written, `permitted_channels.include?(current_channel)` would match any room that's a substring of `permitted_channels`.

Comment (loganmeetsworld) [inline lib/cog_cmd/consul/consul.rb]: 👍 good catch

**Comment by loganmeetsworld** [comment]
@talaris something like this? : 
```
  def run_command
    restrict_command unless permitted_channels.empty?

    key = request.args[0]
    value = request.args[1..-1]
    res = post_key(key, value)

    if success(res)
      response.template = 'default'
      response['body'] = "\n🎉 Value associated with key successfully written."
    else
      raise(Cog::Error, "Error #{res.code}: 💔 Sorry. Your key could not be written.")
    end
  end

  def restrict_command
    raise(Cog::Error, restricted_channels_response) if !permitted_channels.include?(current_channel)
  end
```

**Comment by talaris** [comment]
@loganmeetsworld That seems better, though right now if the ENV var is not set I believe `permitted_channels` would be `nil` not `""`. So you would want to do a `.to_s` first.

This would also impact @ktheory suggestion to split the string into an array. As the split will error if the ENV var is `nil`.

**Review by talaris — APPROVED** [review]
Looks good to me

**Pull request merged**